core/include/cubos/core/metrics.hpp

```diff
@@ -0,0 +1,124 @@
+/// @file
+/// @brief Metrics and profiling utilities and macros.
+/// @ingroup core
+
+#pragma once
+
+#include <algorithm>
+#include <chrono>
+#include <mutex>
+#include <string>
+
+#include <cubos/core/api.hpp>
+#include <cubos/core/reflection/reflect.hpp>
+
+namespace cubos::core
+{
+    /// @brief Represents a metric.
+    struct Metric
+    {
+        CUBOS_REFLECT;
+        std::string mName; ///< The name of the metric.
+        double mValue;     ///< The value of the metric (as a double).
+        std::string mFile; ///< The source file where the metric was created.
+        std::size_t mLine; ///< The line number where the metric was created.
```

Comment:
I don't think this matters for generic metrics. Just the name should be enough.
On the profiler side, you can add this information to the name of the metric - this is necessary as to not have conflicts between functions with the same name.

Reply:
Hmmm, like `myFunction:54:main.cpp` ? The problem will be reading the metric...

Reply:
You would never want to do that though, you would find its name through `readName`.
Btw, we need a way to distinguish profiling metrics from other metrics, maybe we should prefix their name?

Reply:
`prof__myFunction` and `metric__a` looks good?

Reply:
IDK, thinking about it later on the UI, not sure how it should look.

Reply:
On the UI we would probably filter out those and give them special treatment. Maybe we should some special character which doesn't appear in function names? For example, `profiling:<file>:<function name:/line number`

Reply:
so for normal metrics, no prefix, correct?